linter/lints/rfc/lint_cert_via_pkilint.go

```diff
@@ -0,0 +1,152 @@
+	return &certViaPKILint{}
+}
+
+func (l *certViaPKILint) Configure() interface{} {
+	return l
+}
+
+func (l *certViaPKILint) CheckApplies(c *x509.Certificate) bool {
+	// This lint applies to all certificates issued by Boulder, as long as it has
+	// been configured with an address to reach out to. If not, skip it.
+	return l.PKILintAddr != ""
+}
+
+type PKILintResponse struct {
+	Results []struct {
+		Validator           string `json:"validator"`
+		NodePath            string `json:"node_path"`
+		FindingDescriptions []struct {
+			Severity string `json:"severity"`
+			Code     string `json:"code"`
+			Message  string `json:"message,omitempty"`
+		} `json:"finding_descriptions"`
+	} `json:"results"`
+	Linter struct {
+		Name string `json:"name"`
+	} `json:"linter"`
+}
+
+func (l *certViaPKILint) Execute(c *x509.Certificate) *lint.LintResult {
+	timeout := l.PKILintTimeout
+	if timeout == 0 {
+		timeout = 100 * time.Millisecond
+	}
+
+	ctx := context.Background()
+	ctx, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+
+	reqJSON, err := json.Marshal(struct {
+		B64 string `json:"b64"`
+	}{
+		B64: base64.StdEncoding.EncodeToString(c.Raw),
+	})
+	if err != nil {
+		return &lint.LintResult{Status: lint.Error, Details: fmt.Errorf("marshalling pkilint request: %w", err).Error()}
+	}
+
+	url := fmt.Sprintf("%s/certificate/cabf-serverauth", l.PKILintAddr)
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, bytes.NewReader(reqJSON))
+	if err != nil {
+		return &lint.LintResult{
+			Status:  lint.Error,
+			Details: fmt.Errorf("creating pkilint request: %w", err).Error(),
+		}
+	}
+
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return &lint.LintResult{
+			Status:  lint.Error,
+			Details: fmt.Errorf("making POST request to pkilint API: %w", err).Error(),
+		}
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return &lint.LintResult{
+			Status:  lint.Error,
+			Details: fmt.Sprintf("got status %d (%s) from pkilint API", resp.StatusCode, resp.Status),
+		}
+	}
+
+	res, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return &lint.LintResult{
+			Status:  lint.Error,
+			Details: fmt.Errorf("reading response from pkilint API: %w", err).Error(),
+		}
+	}
+
+	var jsonResult PKILintResponse
+	err = json.Unmarshal(res, &jsonResult)
+	if err != nil {
+		return &lint.LintResult{
+			Status:  lint.Error,
+			Details: fmt.Errorf("parsing response from pkilint API: %w", err).Error(),
+		}
+	}
+
+	var findings []string
+	for _, validator := range jsonResult.Results {
+		for _, finding := range validator.FindingDescriptions {
+			id := fmt.Sprintf("%s:%s", validator.Validator, finding.Code)
+			if slices.Contains(l.IgnoreLints, id) {
+				continue
+			}
+			desc := fmt.Sprintf("%s from %s at %s", finding.Severity, id, validator.NodePath)
+			if finding.Message != "" {
+				desc = fmt.Sprintf("%s: %s", desc, finding.Message)
+			}
+			findings = append(findings, desc)
+		}
+	}
+	slices.Sort(findings)
```

Comment:
This sort only seems useful if there are greater than 0 findings and should be moved to line 145. On the other hand, is this sort even necessary at all? In staging/production it would introduce a minor bit of latency during the issuance process, not that that's a terrible thing.

Reply:
Good call on moving the sort a little later.

The sort definitely isn't _necessary_, but due to the structure of the string produced on line 135, this groups them by severity, which is nice for a human reader. I've left a comment to this effect.

Also, sorting a list of a few items is _wildly_ faster than anything else happening in this function. Absolutely dwarfed by the network round-trip, let alone pkilint's parsing and processing.

Reply:
Gotcha. Samantha called me out in a different PR for having a sort that was really only needed in a unit test. Your use case makes sense.